tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts

```diff
@@ -34,6 +43,66 @@ import {Tfdbg2HttpServerDataSource} from '../data_source/tfdbg2_data_source';
 /** @typehack */ import * as _typeHackNgrxStore from '@ngrx/store/src/models';
 /** @typehack */ import * as _typeHackNgrxEffects from '@ngrx/effects/effects';
 
+/**
+ * Getting page indices that are missing from the data and hence need to be
+ * requested.
+ *
+ * @param begin Beginning item index (inclusive).
+ * @param end Ending item index (exclusive).
+ * @param pageSize Size of each page, i.e., how many items there are in each
+ *   complete page. This must be `>= end - begin`, or an Error will be
+ *   thrown.
+ * @param numItems: Total number of items available from the data source.
+ * @param pageLoadedSizes A map from page index to how many items have been
+ *   loaded for that page so far.
+ * @returns An array of the page indices that are currently missing and hence
+ *   should be requested from the appropriate data source.
+ */
+export function getMissingPages(
+  begin: number,
+  end: number,
+  pageSize: number,
+  numItems: number,
+  pageLoadedSizes: {[page: number]: number}
+): number[] {
+  if (pageSize <= 0 || !Number.isInteger(pageSize)) {
+    throw new Error(`Invalid pageSize: ${pageSize}`);
+  }
+  if (end > numItems) {
+    throw new Error(
+      `end index (${end}) exceeds total number of items (${numItems})`
+    );
+  }
+  if (end - begin > pageSize) {
+    throw new Error('begin-end span exceeds page size, which is not allowed');
+  }
+
+  // The constraint that `end - begin <= page` ensures that at most only two
+  // pages need to be requested.
+  const missingPages: number[] = [];
+  const pageIndex0 = Math.floor(begin / pageSize);
+  const anyDigestMissing0 =
+    !(pageIndex0 in pageLoadedSizes) ||
+    (pageLoadedSizes[pageIndex0] < pageSize &&
+      pageIndex0 * pageSize + pageLoadedSizes[pageIndex0] < numItems);
```

Comment:
readability: consider factoring this conditional out and put more descriptive name around the logic.